sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp/client.py

```diff
@@ -895,6 +901,53 @@ def receive_message_batch(self, **kwargs):
         """
         return self._do_retryable_operation(self._receive_message_batch_impl, **kwargs)
 
+    def receive_messages_iter(self, on_message_received=None):
+        """Receive messages by generator. Messages returned in the generator have already been
+        accepted - if you wish to add logic to accept or reject messages based on custom
+        criteria, pass in a callback.
+
+        :param on_message_received: A callback to process messages as they arrive from the
+         service. It takes a single argument, a ~uamqp.message.Message object.
+        :type on_message_received: callable[~uamqp.message.Message]
+        """
+        self._message_received_callback = on_message_received
+        return self._message_generator()
+
+    def _message_generator(self):
+        """Iterate over processed messages in the receive queue.
+
+        :rtype: generator[~uamqp.message.Message]
+        """
+
+        auto_complete = self.auto_complete
+        self.auto_complete = False
+        receiving = True
+        message = None
+        try:
+
+            while receiving:
+
+                receiving = self.do_work()
+
+                message = self._received_messages.get(timeout=self._timeout)
+                if not self._received_messages.empty():
+                    self._received_messages.task_done()
+                yield message
+                
+                self._complete_message(message, auto_complete)
+        except:
+
+            # do we want to close receiver straight out, or as below, only when we have shutdown
+      
+            if self._shutdown:
```

Comment:
this self._shutdown is only set to True when the receiver has hit a value error (client_run) should we just be closing here automatically when we hit the timeout of inactivity. uamqp also did some last message settlement here as this was a finally block